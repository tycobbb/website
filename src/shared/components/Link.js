/* globals SyntheticMouseEvent, SyntheticAnimationEvent, React$Ref */
// @flow
import React, { Component } from 'react'
import { Link as RouteLink } from 'react-router-dom'
import { startsWith } from 'lodash'
import { css } from 'glamor'
import { stylesheet, colors } from '../styles'
import type { RuleProp } from 'glamor'

type State = {
  color: ?string,
  animation: 'stopped' | 'started' | 'ending'
}

export class Link extends Component<*, State> {
  props: {
    to: string | Object,
    unanimated: boolean,
    style?: RuleProp
  }

  static defaultProps = {
    unanimated: false
  }

  state = {
    color: null,
    animation: 'stopped'
  }

  linkRef: ?React$Ref<'a'> = null

  // events
  onMouseEnter = (event: SyntheticMouseEvent<*>) => {
    if (this.props.unanimated) {
      return
    }

    this.setState({ animation: 'started' })
  }

  onMouseLeave = (event: SyntheticMouseEvent<*>) => {
    if (this.props.unanimated) {
      return
    }

    if (this.linkRef) {
      const { color } = window.getComputedStyle(this.linkRef)
      this.setState({ color, animation: 'ending' })
    } else {
      this.setState({ animation: 'stopped' })
    }
  }

  onAnimationEnd = (event: SyntheticAnimationEvent<*>) => {
    if (this.props.unanimated) {
      return
    }

    if (this.state.animation === 'ending') {
      this.setState({ color: null }, () => {
        this.setState({ animation: 'stopped' })
      })
    }
  }

  // helpers
  currentAnimation () {
    switch (this.state.animation) {
      case 'started':
        return rules.animating
      case 'ending':
        return rules.animatingOut
      default: return null
    }
  }

  // lifecycle
  render () {
    const { to, style, ...otherProps } = this.props
    const { color } = this.state

    const rule = css(
      rules.link,
      color && { color },
      this.currentAnimation(),
      style
    )

    const ref = (link) => { this.linkRef = link }
    const props = {
      ...otherProps,
      ...rule,
      onMouseEnter: this.onMouseEnter,
      onMouseLeave: this.onMouseLeave,
      onAnimationEnd: this.onAnimationEnd
    }

    if (typeof to === 'string' && startsWith(to, 'http')) {
      return <a href={to} ref={ref} {...props} />
    } else {
      return <RouteLink to={to} innerRef={ref} {...props} />
    }
  }
}

const animations = {
  rainbow: css.keyframes({
    '0%': { color: colors.purple },
    '8.3%': { color: '#47DAFF' }, // blue
    '24.9%': { color: '#47FF5A' }, // green
    '41.5%': { color: '#FFF547' }, // yellow
    '58.1%': { color: '#FF9A47' }, // orange
    '74.7%': { color: '#FF474A' }, // red
    '100%': { color: colors.purple }
  }),
  rainbowOut: css.keyframes({
    to: { color: colors.purple }
  })
}

const rules = stylesheet({
  link: {
    color: colors.purple
  },
  animating: {
    animation: `${animations.rainbow} 2.0s infinite`
  },
  animatingOut: {
    animation: `${animations.rainbowOut} 0.3s`
  }
})
