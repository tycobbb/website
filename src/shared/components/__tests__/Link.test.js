/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { Link } from '../Link'

let subject

function loadSubject (props) {
  subject = shallow(<Link to='/test' {...props} />)
}

afterEach(() => {
  subject = null
})

describe('#render', () => {
  it('renders properly', () => {
    loadSubject()
    expect(subject).toMatchSnapshot()
  })

  it('renders properly when animating', () => {
    loadSubject()
    subject.setState({ animation: 'started' })
    expect(subject).toMatchSnapshot()
  })

  it('renders properly when ending animation', () => {
    loadSubject()
    subject.setState({ animation: 'ending', color: '#AFAFAF' })
    expect(subject).toMatchSnapshot()
  })
})
