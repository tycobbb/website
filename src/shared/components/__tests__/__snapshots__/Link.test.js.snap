// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render renders properly 1`] = `
<Link
  data-css-d9uk9a=""
  innerRef={[Function]}
  onAnimationEnd={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  replace={false}
  to="/test"
  unanimated={false}
/>
`;

exports[`#render renders properly when animating 1`] = `
<Link
  data-css-gasszt=""
  innerRef={[Function]}
  onAnimationEnd={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  replace={false}
  to="/test"
  unanimated={false}
/>
`;

exports[`#render renders properly when ending animation 1`] = `
<Link
  data-css-vu44jq=""
  innerRef={[Function]}
  onAnimationEnd={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  replace={false}
  to="/test"
  unanimated={false}
/>
`;
